web_services/foxx/brightid5/operations.js

```diff
@@ -0,0 +1,238 @@
+let resetTime = 0;
+function checkLimits(op, timeWindow, limit) {
+  if (Date.now() > resetTime) {
+    operationsCount = {};
+    resetTime =  Date.now() + timeWindow;
+  }
+  const senders = senderAttrs[op.name].map(attr => op[attr]);
+  const usersColl = arango._collection('users');
+  for (let sender of senders) {
+    // these condition structure is applying:
+    // 1) a bucket for a verified user
+    // 2) a bucket for children of a verified user
+    // 3) a bucket for all non-verified users without parent
+    // where parent is the first verified user that make connection with the user
+    if (!usersColl.exists(sender)) {
+      // this happens when operation is "Add Connection" and one/both sides don't exist
+      sender = 'shared';
+    } else {
+      const user = usersColl.document(sender);
+      verified = user.verifications && user.verifications.includes('BrightID');
+      if (!verified && user.parent) {
+        // this happens when user is not verified but has a verified connection
+        sender = `shared_${user.parent}`;
+      } else if (!verified && !user.parent) {
+        // this happens when user is not verified and does not have a verified connection
+        sender = 'shared';
+      }
+    }
+    if (!operationsCount[sender]) {
+      operationsCount[sender] = 0;
+    }
+    operationsCount[sender] += 1;
+    if (operationsCount[sender] <= limit) {
+      // if operation has multiple senders, this check will be passed
+      // even if one of the senders did not reach limit yet
+      return;
+    }
+  }
+  throw 'Too Many Requests';
+}
+
+function verify(op) {
+  if (op.v != 5) {
+    throw 'invalid operation version';
+  }
+  if (op.timestamp > Date.now() + TIME_FUDGE) {
+    throw "timestamp can't be in the future";
+  }
+  const requiredSigs = {
+    'Add Connection': [['id1', 'sig1'], ['id2', 'sig2']],
+    'Remove Connection': [['id1', 'sig1']],
+    'Add Group': [['id1', 'sig1']],
+    'Remove Group': [['id', 'sig']],
+    'Add Membership': [['id', 'sig']],
+    'Remove Membership': [['id', 'sig']],
+    'Set Trusted Connections': [['id', 'sig']],
+    'Set Signing Key': [['id1', 'sig1'], ['id2', 'sig2']],
+    'Sponsor': [['app', 'sig']],
+    'Link ContextId': [['id', 'sig']],
+    'Invite': [['inviter', 'sig']],
+    'Dismiss': [['dismisser', 'sig']],
+    'Add Admin': [['id', 'sig']],
+  }
+  let message = getMessage(op);
+  requiredSigs[op.name].forEach(attrs => {
+    if (op.name != 'Sponsor') {
+      try {
+        verifyUserSig(message, op[attrs[0]], op[attrs[1]]);
+      } catch(e) {
+        // allow adding connections by clients using v4 api
+        // or getting their help to recover
+        // this try and catch should be removed after v4 support dropped
+        if (op.name == 'Add Connection') {
+          const v4message = op.name + op.id1 + op.id2 + op.timestamp;
+          verifyUserSig(v4message, op[attrs[0]], op[attrs[1]]);
+        } else if (op.name == 'Set Signing Key') {
```

Comment:
Why are we only including these operations for v4 deprecated support?  What about the other operations?

Reply:
Because these are the only operations that have multiple signers that can have different version of clients.